include/xtensor/xstridedview.hpp

```diff
@@ -0,0 +1,729 @@
+            {
+                // TODO return empty view instead of throwing error
+                throw std::runtime_error("Offset larger than dim_1 and dim_2");
+            }
+            return diagonal_impl(std::forward<E>(e), offset, axis_1, axis_2, check_policy::none());
+        }
+    }
+
+    /**
+     * @brief Returns a view of the diagonal elements of xexpression e.
+     * If arr has more than two dimensions, then the axes specified by 
+     * axis_1 and axis_2 are used to determine the 2-D sub-array whose 
+     * diagonal is returned. The shape of the resulting array can be 
+     * determined by removing axis_1 and axis_2 and appending an index 
+     * to the right equal to the size of the resulting diagonals.
+     *
+     * @param arr the input array
+     * @param offset offset of the diagonal from the main diagonal. Can
+     *               be positive or negative.
+     * @param axis_1 Axis to be used as the first axis of the 2-D sub-arrays 
+     *               from which the diagonals should be taken. 
+     * @param axis_2 Axis to be used as the second axis of the 2-D sub-arrays 
+     *               from which the diagonals should be taken.
+     * @param check_policy the check level (check_policy::full() or check_policy::none())
+     * @tparam Tag selects the level of error checking on. Default: check_policy::none.
+     * @returns view with values of the diagonal
+     *
+     * \code{.cpp}
+     * xt::xarray<double> a = {{1, 2, 3},
+     *                         {4, 5, 6}
+     *                         {7, 8, 9}};
+     * auto b = xt::diagonal(a); // => {1, 5, 9}
+     * // as this is a view, you can also easily assign new values to 
+     * // the diagonal, e.g.
+     * b = 10;
+     * // => {{ 10,  2,  3}, 
+     * //     {  4, 10,  6},
+     * //     {  7,  8, 10}}
+     * \endcode
+     */
+    template <class E, class Tag = check_policy::none>
+    inline auto diagonal(E&& e, int offset = 0, std::size_t axis_1 = 0, std::size_t axis_2 = 1, Tag check_policy = Tag()) noexcept
+    {
+        return detail::diagonal_impl(std::forward<E>(e), offset, axis_1, axis_2, check_policy);
+    }
+
+    /**
+     * @brief Reverse the order of elements in an xexpression along the given axis.
+     * Note: A NumPy/Matlab style `flipud(arr)` is equivalent to `xt::flip(arr, 0)`,
+     * `fliplr(arr)` to `xt::flip(arr, 1)`.
+     * 
+     * @param arr the input xexpression
+     * @param axis the axis along which elements should be reversed
+     *
+     * @return view evaluating to reversed array
+     */
+    template <class E>
+    inline auto flip(E&& e, std::size_t axis)
+    {
+        using view_type = xstrided_view<xclosure_t<E>>;
+        using shape_type = typename std::decay_t<E>::shape_type;
+        shape_type strides = e.strides();
+        shape_type shape = e.shape();
+        std::size_t data_offset = (shape[axis] - 1) * strides[axis];
+        strides[axis] = -strides[axis];
```

Comment:
yes, i was thinking this too (and wondering why the test cases didn't fail tbh).